### DISCUSSION TITLE
Joint much image tile automatic

### DISCUSSION THREAD

**Discussion by mpsinfo2023** [ImageMagick/ImageMagick#7742] [Digital Image Processing]
Hi, I have to merge many thousands of tile files, obtained from the WEB screen, together. The resolution tile is 4K (4000x1920 pixel).
The files logically have an equal part in common (it is not necessarily always the same overlap size) and the colors of the pixels are exactly identical.
Currently I do this by hand looking for the matching line and joining the 2 images to which I then add the others one at a time. However, this is done if there are few images to be merged.
Can you tell me a solution to use with ImageMagick?

**Comment by snibgo** [comment]
> I have to merge many thousands of tile files, ...

To help us understand what you want, please show some sample input images.

> The files logically have an equal part in common (it is not necessarily always the same overlap size) and the colors of the pixels are exactly identical.

If you know that a certain pair of images have an overlapping portion, then you can make a series of crops around one image, For each crop, search for it in the second image. If you find an exact match, then either this gives the offset for the overlap, or it is a false positive. Ordinary photos are unlikely to give a false positive, if the crops are not too small.

For the merging, I suppose you want to composite one image over the other, with the appropriate offset. For this, "-layers merge" might be used.

**Comment by mpsinfo2023** [comment]
The images always have a part of the overlapping portion, but it is not certain that the offset is always the same, perhaps it is different between the horizontal overlap and the vertical one or even the same between some horizontals in the first row and different in the next row.
Then logically there must be the fusion and therefore the superposition of one photo on the other, unless there is the possibility of finding the same identical row of pixels in common with each other (exactly or even 95-98% if it were to be scanner instead of from the web), crop the images and then place them side by side.

I would also like to know if there is a pixel size limit for the final photo.
Eventually I can also change the format from jpg to tiff or png.
![ZONA_002_001](https://github.com/user-attachments/assets/c4c38f05-d5e8-4dba-bd94-76825d1736ec)
![ZONA_002_002](https://github.com/user-attachments/assets/f3edc073-681e-4a28-a1ed-f7361610b1d2)
![ZONA_002_003](https://github.com/user-attachments/assets/1fba4f25-6f87-46ec-951d-068a27ce2ed7)
![ZONA_002_004](https://github.com/user-attachments/assets/9fd89c40-129e-4d60-9af9-6ff99d7a2934)
![ZONA_002_005](https://github.com/user-attachments/assets/fa864128-e7ee-46ec-b47f-939d5284c59d)
![ZONA_002_006](https://github.com/user-attachments/assets/42122bb7-40eb-480a-9deb-dd20f947e78d)
![ZONA_002_007](https://github.com/user-attachments/assets/e08c0fcd-951b-4ee2-b25c-99f9bd3ef063)
![ZONA_003_001](https://github.com/user-attachments/assets/2784f070-f842-465b-83c4-dbc404c52fca)
![ZONA_003_002](https://github.com/user-attachments/assets/9f420814-2bc5-4676-a961-3cbf759e6e4a)
![ZONA_003_003](https://github.com/user-attachments/assets/b284d1a5-6a50-44ff-9915-ff93ad082174)
![ZONA_003_004](https://github.com/user-attachments/assets/e372c394-a79c-47e4-85ce-0a7be3c8c11a)
![ZONA_003_005](https://github.com/user-attachments/assets/6ed11dc5-24f5-4a34-b979-59281988e919)
![ZONA_003_006](https://github.com/user-attachments/assets/6db617fc-7f16-4520-a986-a6c398cc14f0)
![ZONA_003_007](https://github.com/user-attachments/assets/77f7c241-d27f-4eed-a825-2d58743c0864)
![ZONA_004_001](https://github.com/user-attachments/assets/a123ed97-9077-4f62-9903-935366881d8f)
![ZONA_004_002](https://github.com/user-attachments/assets/af7b0825-f110-4b1f-a225-98e86281d583)
![ZONA_004_003](https://github.com/user-attachments/assets/dcb34f29-5156-487e-a428-1b04670059e2)
![ZONA_004_004](https://github.com/user-attachments/assets/0f0f28f2-2250-48e8-990b-59f1ba9cceed)
![ZONA_004_005](https://github.com/user-attachments/assets/517d333a-bf2f-4eb1-8ddd-68f8fee83b06)
![ZONA_004_006](https://github.com/user-attachments/assets/8b636780-8480-486e-b547-2f22deb8c49e)
![ZONA_004_007](https://github.com/user-attachments/assets/692b55f9-945e-4cf8-9104-30f42a7961c2)

**Comment by fmw42** [comment]
Search Google for file dimension limits. For example: "The maximum width and height of a JPEG image is 65,535 pixels"

**Reply by mpsinfo2023** [reply]
But if I use other formats this limit should be overcome!?... In my archive I have a BMP image with 83,000 pixels on the side.
Which one do you recommend? TIFF or PNG ...

**Comment by snibgo** [comment]
Your images are JPG, which uses lossy compression, so the matches are not exact.

Start with the first two images, which I call m1.jpg and m2.jpg. Make a 10x10 pixel crop from each corner of m2.jpg. Windows BAT syntax.
```
set SIZ=10x10

%IMG7%magick ^
  %SRC2% ^
  ( +clone -gravity NorthWest -crop %SIZ%+0+0 +repage -write c0.png +delete ) ^
  ( +clone -gravity NorthEast -crop %SIZ%+0+0 +repage -write c1.png +delete ) ^
  ( +clone -gravity SouthWest -crop %SIZ%+0+0 +repage -write c2.png +delete ) ^
  ( +clone -gravity SouthEast -crop %SIZ%+0+0 +repage -write c3.png +delete ) ^
  NULL:
```
For each corner, search for it in m1.jpg.
```
magick compare   -similarity-threshold 0 -dissimilarity-threshold 1 -subimage-search   -metric RMSE m1.jpg c0.png NULL:

2379.25 (0.0363051) @ 963,0

magick compare   -similarity-threshold 0 -dissimilarity-threshold 1 -subimage-search   -metric RMSE m1.jpg c1.png NULL:

7496.72 (0.114393) @ 461,89

magick compare   -similarity-threshold 0 -dissimilarity-threshold 1 -subimage-search   -metric RMSE m1.jpg c2.png NULL:

1841.93 (0.0281061) @ 963,420

magick compare   -similarity-threshold 0 -dissimilarity-threshold 1 -subimage-search   -metric RMSEm1.jpg c3.png NULL:

5226.88 (0.0797571) @ 697,281
```
The numbers are a score in the range [0,65535], the same score in the range [0,1], and the x,y coordinates of the top-left of the found match.

The lowest score (best match) is for c2.png, score=0.0281061. This is found at coordinates (963,420). m1.jpg and m2.jpg have height=430, so the found square is at the bottom of the image, at x=963.

So we can merge those two images with:
```
magick m1.jpg ( m2.jpg -page +963+0 ) -layers merge +repage m12.png
```
Repeat that, searching for the corners of the next image in m12.png, and so on.

**Reply by mpsinfo2023** [reply]
Yes, the images sent here are in jpg and therefore compressed due to loading problems, consequently the pixels are not the same, but I can assure you that in my images the pixels match. (file .TIF)
The match percentage thing can still come in handy in the future if the images being merged are not perfectly identical.
To repeat subsequent unions, it is possible to create a loop, how? Alternatively, there is no ready-made code to do an automatic procedure within a folder. I don't have any wait time issues regarding merge execution, as long as the code does an ok job.

**Comment by snibgo** [comment]
I have shown you ImageMagick commands that do what you want. You can put them together in a shell script (eg bash or Windows BAT) to merge any number of files.

My example above does 4 searches for corners, and uses the one with the best (lowest) score. If you know that one search will be exact (score=0), then there is no need to do the other searches.
